[Rachael] iP

rachaelsng opened the pull request:
### Duke

> "Hi, I'm Duke! Nice to meet you!" -[Duke](https://dictionary.cambridge.org/dictionary/english/duke)

Duke helps you keep track of:
- To-dos
- Events
- Deadlines

Here are some other things Duke can do:
1. Save your existing tasks
2. Find a task based on a keyword
3. Find a task based on a date
...and **more**! <3 😉

Let's not forget, Duke is
- [X] Free

_Note:_
```
public static void main(String[] args) {
        new Duke("./data2/tasks.txt").run();
}
```
_The constructor for Duke takes a file path as a parameter. This file path is used to store the task data from Duke._

Commit:
Add Gradle support

rachaelsng committed:
reset

rachaelsng committed:
reset to Level-1

rachaelsng committed:
Level-2

rachaelsng committed:
Level-3

rachaelsng committed:
Level-4

rachaelsng committed:
A-TextUiTesting

rachaelsng committed:
Level-5 (Handled exceptions)

rachaelsng committed:
Level-6 (Delete feature)

rachaelsng committed:
Level-1

rachaelsng committed:
Level-7

ZhengLong3 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,167 @@
+        this.numOfTasks = 0;
+    }
+
+    private String taskTense() {
```
Perhaps your method name should be a verb?

ZhengLong3 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,167 @@
+        duke.greet();
+        boolean isRunning = true;
+
+        while(isRunning) {
```
I think you might be missing a space after the while?

ZhengLong3 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,167 @@
+
+    private void greet() {
+        System.out.println(line);
+        System.out.println("Hi there! I'm Duke\n" +
+                "     What's up?");
```
Maybe it would be more readable if you broke the line before the '+" operator?

ZhengLong3 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,167 @@
+        Duke duke = new Duke();
+        Scanner sc = new Scanner(System.in);
+        duke.greet();
+        boolean isRunning = true;
```
I like how your boolean name sounds like a boolean.

ZhengLong3 reviewed (commented):
Looks good!
Follows coding standards well, with just a few mild violations.

rachaelsng committed:
fix spacing issue

rachaelsng committed:
added dates and times

rachaelsng committed:
handle exceptions for date & time

rachaelsng committed:
Merge branch 'branch-Level-7'

rachaelsng committed:
Merge branch 'branch-Level-8'

rachaelsng committed:
save date & time

rachaelsng committed:
Add command to print tasks on date

sheshenk commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
```
Perhaps all the Task subclasses could be packaged together?

sheshenk commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + dateTime.format(outputFormatter) + ")";
+    }
```
Maybe you could use `String.format` here to make the string concatenation neater?

sheshenk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,232 @@
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.nio.file.*;
+import java.io.*;
```
Could wildcard imports be avoided here, since there aren't many imports from these packages?

sheshenk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,232 @@
+        }
+    }
+
+    private void addTask(String input) throws DukeException {
+        String description = input.substring(input.indexOf(" ") + 1);
+        if (input.startsWith("todo")) {
+            String[] command = input.split("todo", 2);
+            if (command.length < 2 || command[1].trim().length() == 0) {
+                throw new DukeException("Hold up! Description cannot be empty!");
+            }
+            ToDo newTodo = new ToDo(description);
+            tasks.add(numOfTasks, newTodo);
+            numOfTasks++;
+        } else if (input.startsWith("deadline")) {
+            String[] command = input.split("deadline", 2);
+            String[] date = command[1].split("/by", 2);
+            if (command[1].trim().length() == 0 || date[0].trim().length() == 0) {
+                throw new DukeException("Hold up! Description cannot be empty!");
+            }
+            if (date.length < 2 || date[1].trim().length() == 0) {
+                throw new DukeException("Wait! When do you want to do this by??");
+            }
+            Deadline newDeadline = new Deadline(description.substring(0, description.indexOf("/by") - 1),
+                    description.substring(description.indexOf("/") + 4));
+            tasks.add(numOfTasks, newDeadline);
+            numOfTasks++;
+        } else if (input.startsWith("event")) {
```
This method is quite long. Perhaps you could have other helper methods to reduce the number of lines in this method? Creating `Ui` and `Parser` class abstractions might help!

sheshenk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,232 @@
+        Path p = Paths.get("./data/duke.txt");
+        Path dataPath = Paths.get("./data");
+
+        boolean isDataFileExisting = Files.exists(dataPath);
```
I like that your booleans sound like booleans!

sheshenk commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,232 @@
+            while ((line = reader.readLine()) != null) {
+                Task newTask;
+                // process data
+                switch (line.charAt(4)) {
```
Could the `4` be a static variable? This could reduce the reliance on 'magic numbers'

sheshenk commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,33 @@
+        return new Event(line.substring(10, line.indexOf("(at:") - 1),
+                LocalDateTime.parse(line.substring(line.indexOf("(at:") + 5, line.lastIndexOf(")")),
+                        outputFormatter));
+    }
```
Could the individual parameters be aligned? You could try using a formatting similar to the Egyptian braces when you have many parameters for a function.

sheshenk reviewed (commented):
Overall, this is a good PR! The structure of the application can be improved by packaging the various classes. The quality and standard of the code is good as well!

rachaelsng committed:
Add Ui, Storage, Parser, Command & TaskList classes

rachaelsng committed:
Put all classes into duke package

rachaelsng committed:
Divide into task and commands packages

rachaelsng committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

rachaelsng committed:
Added some JUnit tests

rachaelsng committed:
Fixed mainCLassName in build.gradle

rachaelsng committed:
Added JavaDocs for Ui & TaskList

rachaelsng committed:
Added JavaDocs for Command & Parser, added fullstops for JavaDocs

rachaelsng committed:
Added JavaDocs for commands

rachaelsng committed:
Added JavaDocs for tasks, Duke & DukeException

rachaelsng committed:
Adhere to coding standards

rachaelsng committed:
Set class variables to private

rachaelsng committed:
Add find functionality

rachaelsng committed:
Merge branch 'branch-A-JavaDoc'

rachaelsng committed:
Merge branch 'branch-A-CodingStandard and resolve conflicts

rachaelsng committed:
Merge branch 'branch-Level-9' and resolve conflicts

rachaelsng committed:
Fix Storage bug

rachaelsng committed:
Fix checkstyle violations

rachaelsng committed:
Edit .gitignore

rachaelsng committed:
Add config files

rachaelsng committed:
Add GUI using JavaFX

rachaelsng committed:
Merge branch 'branch-Level-10'

rachaelsng committed:
Change Duke to Gina

Chatbot is the same as the given example.

A more unique chatbot personality would be more engaging for users.

Let's change
* the dialogue from the chatbot
* the display pictures for the chatbot and user

Doing so will represent Gina's personality better.

rachaelsng committed:
Update build.gradle application and shadowJar configurations

rachaelsng committed:
Add assertions

No assertions are used.

There may be undetected bugs in the code.

Let's use the assert feature to document important assumptions that should hold at various points in the code.

Assertions are used to define assumptions about the program state so that the runtime can verify them.

rachaelsng committed:
Modify code to improve code quality

rachaelsng committed:
Merge pull request #2 from rachaelsng/branch-A-Assertions

Add assertions

rachaelsng committed:
Merge pull request #3 from rachaelsng/branch-A-CodeQuality

Modify code to improve code quality

rachaelsng committed:
Merge branch 'master' into branch-A-CodeQuality

rachaelsng committed:
Add contacts and update tests

rachaelsng committed:
Add files under version control

rachaelsng committed:
Add bye functionality

rachaelsng committed:
Update README.md

rachaelsng committed:
Modify README.md

rachaelsng committed:
Update JavaDocs

rachaelsng committed:
Add Ui.png

rachaelsng committed:
Handle contact with no info
